File: libsrc/gather.f90
module gather_module

use, intrinsic :: iso_fortran_env, only : real64
implicit none

contains

subroutine collect_obs_simul(srcxyz,m,gather,ngather)
!observed arrivals, weights and
!calculated travel times, dt/dx for all stations

use arrivals,           only: parrs  !, cFileArr, cRecordArr
use twopoint

implicit none

real,dimension(3), intent(IN)       :: srcxyz
integer, intent(IN)                 :: m
!gather(1,:) calc tt
!gather(2,:) calc dt/dx
!gather(3,:) calc dt/dy
!gather(4,:) calc dt/dz
!gather(5,:) obs arrival
!gather(6,:) obs weight
real(real64),dimension(7,m),intent(OUT) :: gather
integer, intent(OUT)                    :: ngather

integer           :: i,j,narr
real,dimension(3) :: obsxyz
character(LEN=5)  :: ph
real,dimension(4) :: td
real              :: toa

narr=parrs%n_arr
if (narr .gt. m) then
    narr=m
end if
j=0
do i=1,narr
    obsxyz(1)=parrs%arr(i)%X
    obsxyz(2)=parrs%arr(i)%Y
    obsxyz(3)=parrs%arr(i)%Z
    ph=trim(parrs%arr(i)%phase)
    call two_point_td(ph,srcxyz,obsxyz,td,toa)
    !if (td(1) .le. 0.0) cycle
    j=j+1
    !write(*,*) ph,srcxyz,obsxyz,td,toa
    !tcal(i)=td(1)+dly(key(i))  !  travel time
    gather(1,j)=td(1)              !  travel time
    gather(2,j)=td(2)              !  dt/dx
    gather(3,j)=td(3)              !  dt/dy
    gather(4,j)=td(4)              !  dt/dz
    !  dt/dt_o = 1
    if (td(1) .gt. 0.0) then
        gather(5,j)=parrs%arr(i)%trec
        gather(6,j)=parrs%arr(i)%wt
        gather(7,j)=toa
    else
        gather(5,j)=parrs%arr(i)%trec
        gather(6,j)=0.0
        gather(7,j)=toa
    end if
end do
ngather=j
end subroutine collect_obs_simul

end module gather_module

File: libsrc/gather.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

#include "arrivals.h"
#include "twopoint.h"
/*
// Gather class
// This class is used to collect the travel time information from the arrivals
// and calculate the residuals and weights for the inversion process.

class Gather {
private:
   int refresh_token = 0;
public:
   int ngather = 0; // number of gathers
   std::vector<double> ttcal;  // travel time calculated
   std::vector<double> dtdx;   // dt/dx
   std::vector<double> dtdy;   // dt/dy
   std::vector<double> dtdz;   // dt/dz
   std::vector<double> toa;    // take-off angle
   std::vector<double> trec;   // time recorded 
   std::vector<double> wt;     // weight

   void resize(int n) {
       ttcal.resize(n);
       dtdx.resize(n);
       dtdy.resize(n);
       dtdz.resize(n);
       toa.resize(n);
       trec.resize(n);
       wt.resize(n);
       ngather = n;
   }  

  // ~Gather() {
  // }  

   void clear() {
	ttcal.clear();
	dtdx.clear();
	dtdy.clear();
	dtdz.clear();
	toa.clear();
	trec.clear();
	wt.clear();
        ngather = 0;
   }

};
*/
#include "gather.h"

//   int Gather::collect(const double (& x)[4]) {

   int Gather::collect(const double* x) {
       std::string ph;
       double td[4];
       double takeoffangle;
       double obsxyz[3];
       double srcxyz[3] = {x[0], x[1], x[2]};

       int narr = arrs.arr.size();
       if (narr == 0) {
	   std::cerr << "No arrivals in the list" << std::endl;
	   return 0;
       }
       if (narr != ngather) {
	   if (ngather != 0) {
	       clear();
	       std::cerr << "Clearing the gather" << std::endl;
	   }
	   resize(narr);
       }
       auto a = arrs.arr;
       int i;
       for (i = 0; i < narr; ++i) {
	   obsxyz[0] = a[i].X;
	   obsxyz[1] = a[i].Y;
	   obsxyz[2] = a[i].Z;
	   ph = a[i].phase;
	   two_point_td(srcxyz, obsxyz, ph, td, takeoffangle);
	   
	   ttcal[i] = td[0]; // travel time
	   dtdx[i]  = td[1]; // dt/dx
	   dtdy[i]  = td[2]; // dt/dy
	   dtdz[i]  = td[3]; // dt/dz
	   toa[i]   = takeoffangle;   // take-off angle
	   
	   trec[i] = a[i].trec;  // time recorded
	   
	   if (td[0] > 0.0) {
	   // weight
	       wt[i] = a[i].wt;
	   } else {
	       wt[i] = 0.0;
	   }
       }
       return i;
   }


//   int Gather::get_topt(double(& x)[4]) {

   int Gather::get_topt(double* x) {

        //collect({x[0], x[1], x[2]});

        double numerator = 0.0;
        double denominator = 0.0;
	int j;
        for (j = 0; j < ngather; j++) {
            numerator += (trec[j] - ttcal[j]) * wt[j];
            denominator += wt[j];
        }
        if (denominator != 0.0) {
            x[3] = numerator / denominator;
        } else {
            x[3] = 0.0;
        }
	return j;
    }

//    void Gather::get_res_w(const double(& x)[4], double* fvec, double* w) {
//        collect({x[0], x[1], x[2]});
      
//    int Gather::get_res_w(const double(& x)[4], int ldjac, double* fvec, double* w) {

    int Gather::get_res_w(const double* x, int ldjac, double* fvec, double* w) {
	int m = std::min(ldjac, ngather);
	int j;
        for (j = 0; j < m; j++) {
            fvec[j] = trec[j] - x[3] - ttcal[j];
            w[j] = wt[j];
        }
	return j;
    }

//    void Gather::get_G_w(const double (& x)[4], double fjac[][100], double* w) {
//        collect({x[0], x[1], x[2]});
        
    int Gather::get_G_w(int ldjac, double* fjac, double* w) {
	int m = std::min(ldjac, ngather);
	int j;
        for (j = 0; j < m; j++) {
            fjac[j+0*ldjac] = -dtdx[j];
            fjac[j+1*ldjac] = -dtdy[j];
            fjac[j+2*ldjac] = -dtdz[j];
            fjac[j+3*ldjac] = -1.0;
            w[j] = wt[j];
        }
	return j;
    }

Gather gather;
